feat: support configureable sorting mode for imports, keys and attributes

nazarhussain opened the pull request:
## Summary

Add a configuration for the formatter `sortMode` with either `alphabetical` or `natural` (default) options. This sorting mode applies to the import sorting, sorting keys and sorting attributes. 

## Motivation 

Not all projects and teams have same preferences when it comes to code formatting, and it's much inclusive approach to give them option rather than enforcing one format. In addition:

- Natural sorting seems good mostly when there are numbers involved in the names of literals. My personal observation is that's not a big use case among many projects and codebases. 
- When it comes to projects which have mix of types and code, mostly the types follow PascalCase and others follow either camelCase or CAPITAL_CASE, in that format alphabetical sorting seems more readable. 


## Test Plan

Added the spec tests

**NOTE**: 

- I am not a Rust guru, so there may be some points of improvements for performance. 
- I wanted to keep the sorting mode singletone and cloned for different file formats but could not find better way in current code structure.

dyc3 commented:
Thanks for the PR! The scope of this PR is pretty big. IMO, I would prefer if a discussion was opened first so that we could gather community feedback on the feature to see if there's enough demand for it, and to agree and align on scope.

I'm personally on the fence about how useful this would realistically be -- the value add feels negligible.

Implementation wise, it's a bit unusual to add a formatter option where it's mostly used in assists. We have a pretty high bar for adding formatter options. It would make a lot more sense (at least for now) to have the option be per-rule.

nazarhussain commented:
> I'm personally on the fence about how useful this would realistically be -- the value add feels negligible.

I feel that value addition is particularly based on the nature of project. The moment I upgrade Biome, it was a clear that my team will not go along this kind of import sorting. And there is also a demand building up from the community https://github.com/biomejs/biome/discussions/6122, and from that you can see different impact of selecting sorting logic. If you feel I can open open discussion in Github/Discord anywhere you want to gather more feedback. 

> Implementation wise, it's a bit unusual to add a formatter option where it's mostly used in assists.

Yes, I started PR earlier by setting up option per assist, but later observed that this is actually a formatter option like `indent_style`, `attribute_position` and `bracket_spacing` etc this seems more natural that `sort_mode` come under formatter. If you prefer otherwise, I can move those to per assist option.

matthewkeil commented:
Signaling support for this feature 🚀

arendjr commented:
So I think for now you will have to duplicate the options. Would that work for you?

dyc3 commented:
I don't think so. The problem comes from there being 2 rules, one for each language, so the name conflicts.

nazarhussain commented:
> I don't think so. The problem comes from there being 2 rules, one for each language, so the name conflicts.

Yes two same name rules for two languages. One for json and other for js, but from configuration exported as json options. 

The assert logic want it to be exact same reference of struct. So it works for json but not for js. 

Good approach would be to assert if the options are same not the exact object reference.

nazarhussain commented:
I tried to only have methods inside the compareable_token now in reference to the not put the sort_mode into every token, that was redundant and overwhelming for memory. 

Please review now @dyc3 @arendjr 

Related to the error for cross crate options for similar rules, it seems to have following options for us. 

1. Add simple crate for it in this PR starting with options for `useSortedKeys` for JS and JSON, and then add more in next PRs for other options. 
2. We already using JSON-Schema for options, so we may implement schema fingerprint hashing, and then compare those. As long options are same, the schema for options will be same and that will work in comparison. 

Please review the code and suggest the next steps.

arendjr commented:
It seems https://github.com/biomejs/biome/pull/5543 is almost ready to be merged. Is it okay we keep this on hold until it is ready? You'll probably need a rebase after that, but it should solve the issues with the codegen/schema.

Conaclos commented:
Some questions:

- What "alphabetic" order do you want for non-ASCII characters?
- Regarding `orgainizeImports`, the sort order only affects ordering of named specifiers and import attributes, not the import source?
- Why did you name it `sortMode`? Any rationales? Have you some possible alternative names?
-

Conaclos commented:
> Strings are ordered lexicographically by their byte values. This orders Unicode code points based on their positions in the code charts. This is not necessarily the same as "alphabetical" order, which varies by language and locale.

We have a little difference in our side: our order if byte-based while their order is codepoint-based.
I left others leave their opinion.
I have no strong preference here. `lexicographic` could be ok.

arendjr commented:
> We have a little difference in our side: our order if byte-based while their order is codepoint-based.

I don't think that's true. Rust's ordering is byte-based too.

Anyway, I think `lexicographic` has my preference too as `binary` sounds a bit technical indeed. Otherwise happy with the choices made :+1:

Conaclos commented:
> I don't think that's true. Rust's ordering is byte-based too.

Actually I am wrong.
Indeed, [it is bytewise](https://doc.rust-lang.org/src/core/str/traits.rs.html#20).
Let's stick with `lexicographic`.

nazarhussain commented:
@Conaclos @arendjr To me it seems all pints are covered. Please suggest if there is any thing further needed to be taken care of.

Conaclos commented on crates/biome_rule_options/src/shared/sort_order.rs:
```diff
@@ -0,0 +1,9 @@
+    Clone, Copy, Debug, Default, Eq, PartialEq, serde::Deserialize, serde::Serialize, biome_deserialize_macros::Deserializable,
+)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+pub enum SortOrder {
```
I think you should add:

```suggestion
#[serde(rename_all = "camelCase")]
pub enum SortOrder {
```

nazarhussain replied:
Resolved

Conaclos commented on crates/biome_string_case/src/lib.rs:
```diff
@@ -672,6 +677,10 @@ impl StrLikeExtension for str {
         self.as_bytes().ascii_nat_cmp(other.as_bytes())
     }
+
+    fn lexicographic_cmp(&self, other: &Self) -> Ordering {
+        self.as_bytes().lexicographic_cmp(other.as_bytes())
+    }
```
How is this different from Rust's built-in `cmp`?

Conaclos commented on crates/biome_string_case/src/lib.rs:
```diff
@@ -704,6 +713,10 @@ impl StrLikeExtension for std::ffi::OsStr {
             .ascii_nat_cmp(other.as_encoded_bytes())
     }
+
+    fn lexicographic_cmp(&self, other: &Self) -> Ordering {
+        self.as_encoded_bytes().cmp(other.as_encoded_bytes())
+    }
```
How is this different from Rust's built-in `cmp`?

nazarhussain replied:
As mentioned above.

Conaclos commented on crates/biome_string_case/src/lib.rs:
```diff
@@ -719,6 +732,10 @@ impl StrLikeExtension for [u8] {
         CldrAsciiCollator.cmp(self.iter().copied(), other.iter().copied())
     }
+
+    fn lexicographic_cmp(&self, other: &Self) -> Ordering {
+        self.iter().copied().cmp(other.iter().copied())
+    }
```
How is this different from Rust's built-in `cmp`?

nazarhussain commented:
@arendjr I had fixed the lint/fmt issues and also added the changeset. For the website update, is it possible to open a PR there now or have to wait for this one to be merged first? The [instructions](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#documentation) are not that much clear for understanding.

nazarhussain commented:
@ematipico Thanks for your feedback. There were no invalid/valid cases for the assists I believe so I added examples which shows what diagnostic after sorting. 

Let me know if that's good or suggest any changes.
